tools/bridge/bridge/transfer_fetcher.py

```diff
@@ -1,25 +1,135 @@
 import logging
-
 import gevent
-from gevent.queue import Queue
 
+from gevent.queue import Queue
 from web3 import Web3
 
+MINIMAL_TOKEN_ABI = [
+    {
+        "anonymous": False,
+        "inputs": [
+            {"indexed": True, "name": "from", "type": "address"},
+            {"indexed": True, "name": "to", "type": "address"},
+            {"indexed": False, "name": "value", "type": "uint256"},
+        ],
+        "name": "Transfer",
+        "type": "event",
+    }
+]
+
+TRANSFER_EVENT_SIGNATURE_HASH = Web3.keccak(text="Transfer(address,address,uint256)")
+
+
+class TransferFetcher:
+    def __init__(
+        self,
+        *,
+        queue: Queue,
+        w3_foreign: Web3,
+        token_contract_address: str,
+        foreign_bridge_contract_address: str,
+        max_reorg_depth: int,
+        transfer_event_fetch_limit: int,
+    ):
+        token_contract_code = w3_foreign.eth.getCode(token_contract_address)
+
+        if not token_contract_code:
+            raise ValueError(
+                f"The given token contract address {token_contract_address} "
+                "does not point to a contract!"
+            )
+
+        if TRANSFER_EVENT_SIGNATURE_HASH not in token_contract_code:
+            raise ValueError(
+                f"The token contract at the given address {token_contract_address}"
+                "does not have the (correct) 'Transfer' event!"
+            )
+
+        foreign_bridge_contract_code = w3_foreign.eth.getCode(
+            foreign_bridge_contract_address
+        )
+
+        if not foreign_bridge_contract_code:
+            raise ValueError(
+                f"The given foreign bridge contract address {foreign_bridge_contract_address} "
+                "does not point to a contract!"
+            )
+
+        self._logger = logging.getLogger("bridge.transfer_fetcher")
+        self._w3_foreign = w3_foreign
+        self._queue = queue
+        self._token_contract = w3_foreign.eth.contract(
+            address=token_contract_address, abi=MINIMAL_TOKEN_ABI
+        )
+        self._foreign_bridge_contract_address = foreign_bridge_contract_address
+        self._max_reorg_depth = max_reorg_depth
+        self._transfer_event_fetch_limit = transfer_event_fetch_limit
+        self._last_fetched_block_number = (
```

Comment:
can we get rid of this and just keep it as a variable in the method and pass it as a parameter? Less state, easier to understand

Reply:
So you would give it as argument and return it back? :thinking:

Reply:
No, I would give both start and end block number as an argument. Next time you give `end` as start, and so on. Then

Reply:
This is what I'm doing. The storage of `last_fetched_block_number` is for between the calls of `fetch_events_not_seen`. So he knows where he stopped last time. Then he is generating the `from_block_number` and `to_block_number` values for blocks which haven't been seen respecting the maximum reorg depth and also the event limits per RPC request.
Please call me if this is still unclear or if I just don't understand you. Probably the last case.